plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteVectorDistanceFunction.java

```diff
@@ -0,0 +1,159 @@
+                .with(argument(0).matching(expression().capturedAs(LEFT_ARGUMENT).with(type().matching(RewriteVectorDistanceFunction::isArrayTypeWithRealOrDouble))))
+                .with(argument(1).matching(expression().capturedAs(RIGHT_ARGUMENT).with(type().matching(RewriteVectorDistanceFunction::isArrayTypeWithRealOrDouble))));
+        this.operator = requireNonNull(operator, "operator is null");
+    }
+
+    @Override
+    public Pattern<? extends ConnectorExpression> getPattern()
+    {
+        return pattern;
+    }
+
+    @Override
+    public Optional<JdbcExpression> rewrite(ConnectorExpression projectionExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Optional<ParameterizedExpression> leftExpression = rewrite(captures.get(LEFT_ARGUMENT), context);
+        if (leftExpression.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<ParameterizedExpression> rightExpression = rewrite(captures.get(RIGHT_ARGUMENT), context);
+        if (rightExpression.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new JdbcExpression(
+                "%s %s %s".formatted(leftExpression.get().expression(), operator, rightExpression.get().expression()),
+                ImmutableList.<QueryParameter>builder()
+                        .addAll(leftExpression.get().parameters())
+                        .addAll(rightExpression.get().parameters())
+                        .build(),
+                new JdbcTypeHandle(
+                        Types.DOUBLE,
+                        Optional.of("double"),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty())));
+    }
+
+    private static Optional<ParameterizedExpression> rewrite(ConnectorExpression expression, RewriteContext<ParameterizedExpression> context)
+    {
+        if (expression instanceof Constant constant) {
+            Object value = constant.getValue();
+            LongArrayBlock block = (LongArrayBlock) value;
+            StringBuilder vector = new StringBuilder("'[");
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    return Optional.empty();
+                }
+                double doubleValue = longBitsToDouble(block.getLong(i));
+                if (!isSupportedVector(doubleValue)) {
+                    return Optional.empty();
+                }
+                vector.append(doubleValue);
+                if (i != block.getPositionCount() - 1) {
+                    vector.append(",");
+                }
+            }
+            vector.append("]'");
+            return Optional.of(new ParameterizedExpression(vector.toString(), ImmutableList.of()));
+        }
+        if (expression instanceof Call call && call.getFunctionName().equals(CAST_FUNCTION_NAME)) {
```

Comment:
Shouldn't `CAST` handled by a different rewriter ?

Reply:
That was my initial implementation, but merged into one to simplify the code. Is there a situation we can't handle without a separated `CAST` rewriter?

Reply:
Can we have an abstract rule with two implementation - one for double and real (For real we could have `CAST` as a part of the pattern right ?